### PR TITLE
Conversion of classes extending from EventHandler to ES6

### PR DESCRIPTION
**Pull request by mvaligursky**
Leftover files to convert:

- [x] script-type should be done in this PR as well, extends EventHandler

### CODE REVIEW

File: src/asset/asset-list-loader.js [hunk 1]
```diff
@@ -11,188 +11,187 @@ import { Asset } from './asset.js';
...
-            done.call(scope, assets);
-        });
+    /**
+     * @private
+     * @function
+     * @name pc.AssetListLoader#ready
+     * @param {Function} done - Callback called when all assets in the list are loaded.
+     * @param {object} [scope] - Scope to use when calling callback.
+     */
+    ready = function (done, scope) {
```
Comment (willeastcott) [inline src/asset/asset-list-loader.js:100]:
```suggestion
    ready(done, scope) {
```

File: src/core/tags.js [hunk 2]
```diff
@@ -1,26 +1,26 @@
 import { EventHandler } from './event-handler.js';
 
-function TagsCache(key) {
-    this._index = { };
-    this._key = key || null;
-}
+class TagsCache {
+    constructor(key) {
```
Comment (willeastcott) [inline src/core/tags.js:4]:
```suggestion
    constructor(key = null) {
```

File: src/core/tags.js [hunk 3]
```diff
@@ -1,26 +1,26 @@
...
+        this._index = { };
+        this._key = key || null;
```
Comment (willeastcott) [inline src/core/tags.js:6]:
```suggestion
        this._key = key;
```

File: src/framework/components/sprite/component.js [hunk 4]
```diff
@@ -389,20 +387,20 @@ Object.assign(SpriteComponent.prototype, {
...
         // pivot
         this._node.setLocalPosition(posX, posY, 0);
-    },
+    }
 
     // updates AABB while 9-slicing
-    _updateAabb: function (aabb) {
-        // pivot
+    _updateAabb(aabb) {
+            // pivot
```
Comment (willeastcott) [inline src/framework/components/sprite/component.js:394]:
```suggestion
        // pivot
```

File: src/scene/graph-node.js [hunk 5]
```diff
@@ -14,6 +14,13 @@ var scaleCompensateScale = new Vec3();
 var scaleCompensateScaleForParent = new Vec3();
 var tmpMat4 = new Mat4();
 var tmpQuat = new Quat();
+var position = new Vec3();
+var invParentWtm = new Mat4();
+var rotation = new Quat();
+var invParentRot = new Quat();
+var matrix = new Mat4();
+var target = new Vec3();
+var up = new Vec3();
```
Comment (willeastcott) [inline src/scene/graph-node.js:23]: We can probably merge some of these globals into a minimal set of temp vec/quat/mats.

File: src/scene/scene.js [hunk 6]
```diff
@@ -90,496 +90,485 @@ import { Mat4 } from '../math/mat4.js';
...
+     * @name pc.Scene#setSkybox
+     * @description Sets the cubemap for the scene skybox.
+     * @param {pc.Texture[]} [cubemaps] - An array of cubemaps corresponding to the skybox at different mip levels. If undefined, scene will remove skybox.
+     * Cubemap array should be of size 7, with the first element (index 0) corresponding to the base cubemap (mip level 0) with original resolution.
+     * Each remaining element (index 1-6) corresponds to a fixed prefiltered resolution (128x128, 64x64, 32x32, 16x16, 8x8, 4x4).
+     */
+    setSkybox(cubemaps) {
+        var i;
+        if (!cubemaps)
+            cubemaps = [null, null, null, null, null, null, null];
```
Comment (willeastcott) [inline src/scene/scene.js:470]: Use default parameter?

File: src/scene/sprite.js [hunk 7]
```diff
@@ -51,220 +51,219 @@ var spriteIndices = [
...
- * @name pc.Sprite#destroy
- * @description Free up the meshes created by the sprite.
- */
-Sprite.prototype.destroy = function () {
+    /**
+     * @function
+     * @name pc.Sprite#destroy
+     * @description Free up the meshes created by the sprite.
+     */
+    destroy = function () {
```
Comment (willeastcott) [inline src/scene/sprite.js:253]:
```suggestion
    destroy() {
```

File: src/scene/texture-atlas.js [hunk 8]
```diff
@@ -27,93 +27,91 @@ import { EventHandler } from '../core/event-handler.js';
...
-    if (this._texture) {
-        this._texture.destroy();
+    /**
+     * @function
+     * @name pc.TextureAtlas#removeFrame
+     * @param {string} key - The key of the frame.
+     * @example
+     * atlas.removeFrame('1');
+     */
+    removeFrame = function (key) {
```
Comment (willeastcott) [inline src/scene/texture-atlas.js:79]:
```suggestion
    removeFrame(key) {
```

File: src/scene/texture-atlas.js [hunk 9]
```diff
@@ -27,93 +27,91 @@ import { EventHandler } from '../core/event-handler.js';
...
+            this.fire('remove:frame', key.toString(), frame);
+        }
+    }
+
+    /**
+     * @function
+     * @name pc.TextureAtlas#destroy
+     * @description Free up the underlying texture owned by the atlas.
+     */
+    destroy = function () {
```
Comment (willeastcott) [inline src/scene/texture-atlas.js:92]:
```suggestion
    destroy() {
```

File: src/asset/asset-registry.js [hunk 10]
```diff
@@ -18,159 +18,157 @@ import { Asset } from './asset.js';
...
-Object.assign(AssetRegistry.prototype, {
     /**
      * @function
      * @name pc.AssetRegistry#list
      * @description Create a filtered list of assets from the registry.
      * @param {object} filters - Properties to filter on, currently supports: 'preload: true|false'.
      * @returns {pc.Asset[]} The filtered list of assets.
      */
-    list: function (filters) {
+    list(filters) {
```
Comment (willeastcott) [inline src/asset/asset-registry.js:171]: Use default parameter here.

File: src/asset/asset.js [hunk 11]
```diff
@@ -70,89 +70,87 @@ var VARIANT_DEFAULT_PRIORITY = ['pvr', 'dxt', 'etc2', 'etc1', 'basis'];
...
+        super();
 
-    this._id = assetIdCounter--;
+        this._id = assetIdCounter--;
 
-    this.name = name || '';
-    this.type = type;
-    this.tags = new Tags(this);
-    this._preload = false;
+        this.name = name || '';
```
Comment (willeastcott) [inline src/asset/asset.js:79]: Use default parameter.

Comment (willeastcott) [reply src/asset/asset.js:79]: Although having said that, the docs don't say the `name` parameter is optional. So maybe don't.

File: src/asset/asset.js [hunk 12]
```diff
@@ -70,89 +70,87 @@ var VARIANT_DEFAULT_PRIORITY = ['pvr', 'dxt', 'etc2', 'etc1', 'basis'];
...
 
-    this.variants = new AssetVariants(this);
+        this.variants = new AssetVariants(this);
 
-    this._file = null;
-    this._data = data || { };
-    this.options = options || { };
+        this._file = null;
+        this._data = data || { };
+        this.options = options || { };
```
Comment (willeastcott) [inline src/asset/asset.js:88]: These two can be default parameters though.

File: src/framework/components/anim/controller.js [hunk 13]
```diff
@@ -20,67 +20,74 @@ import {
...
-    this._weight = 1.0;
-    this._animTrack = null;
-}
+class AnimNode {
+    constructor(state, parent, name, point, speed) {
+        this._state = state;
+        this._parent = parent;
+        this._name = name;
+        this._point = Array.isArray(point) ? new pc.Vec2(point) : point;
+        this._speed = speed || 1.0;
```
Comment (willeastcott) [inline src/framework/components/anim/controller.js:29]: Make default parameter. Also, JSDoc above it should be updated to specify `speed` is optional, defaulting to 1.

Comment (willeastcott) [reply src/framework/components/anim/controller.js:29]: There's also a doc bug for point - it's an array of numbers.

Comment (willeastcott) [reply src/framework/components/anim/controller.js]: Fixed.

**Review by willeastcott — APPROVED** [review]

File: src/asset/asset.js (hunk 12 above)
Comment (mvaligursky) [reply src/asset/asset.js]: done

**Comment by willeastcott** [comment]
I've run through all the engine examples with this branch. All looks good. I'm happy for it to be merged. The only [bug](https://github.com/playcanvas/engine/issues/2695) I found was unrelated to this PR.

**Comment by willeastcott** [comment]
I've migrated ScriptType. 😄

**Pull request merged**